File: job-portal-app/src/app/core/services/job.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface Company {
  companyId: number;
  companyName: string;
  logoUrl?: string;
}

export interface Job {
  jobId: number;
  companyId: number;
  postedByUserId: number;
  categoryId: number;

  title: string;
  jobType: string;
  workMode: string;
  location: string;

  salaryMin?: number;
  salaryMax?: number;

  description: string;
  responsibilities?: string;
  requirements?: string;
  benefits?: string;

  deadline?: string;
  status?: string;

  createdAt?: string;
  updatedAt?: string;

  company?: Company;
}

export interface CreateJobDto {
  companyId: number;
  postedByUserId: number;
  categoryId: number;

  title: string;
  jobType: string;
  workMode: string;
  location: string;

  salaryMin?: number;
  salaryMax?: number;

  description: string;
  responsibilities?: string;
  requirements?: string;
  benefits?: string;

  deadline?: string;
  status?: string;
}

export interface UpdateJobDto {
  companyId: number;
  postedByUserId: number;
  categoryId: number;

  title: string;
  jobType: string;
  workMode: string;
  location: string;

  salaryMin?: number;
  salaryMax?: number;

  description: string;
  responsibilities?: string;
  requirements?: string;
  benefits?: string;

  deadline?: string;
  status?: string;
}


export interface PaginatedResult<T> {
  items: T[];
  totalCount: number;
  page: number;
  pageSize: number;
}

@Injectable({
  providedIn: 'root'
})
export class JobService {

  private http = inject(HttpClient);

  // change this to your API url
  //private apiUrl = 'https://localhost:7168/api/jobs';
  private readonly apiUrl = environment.baseUrl+'jobs';

  /**
   * Get all jobs
   */
  getJobs(): Observable<Job[]> {
    return this.http.get<Job[]>(this.apiUrl);
  }

  /**
   * Get single job by id
   */
  getJobById(id: number): Observable<Job> {
    return this.http.get<Job>(`${this.apiUrl}/${id}`);
  }

  /**
   * Get paginated jobs
   */
  getJobsPaginated(
    page: number = 1,
    pageSize: number = 10
  ): Observable<PaginatedResult<Job>> {

    const params = new HttpParams()
      .set('page', page)
      .set('pageSize', pageSize);

    return this.http.get<PaginatedResult<Job>>(
      `${this.apiUrl}/paginated`,
      { params }
    );
  }

  /**
   * Create new job
   */
  createJob(data: CreateJobDto): Observable<Job> {
    return this.http.post<Job>(this.apiUrl, data);
  }

  /**
   * Update existing job
   */
  updateJob(id: number, data: UpdateJobDto): Observable<void> {
    return this.http.put<void>(`${this.apiUrl}/${id}`, data);
  }

  /**
   * Delete job
   */
  deleteJob(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  /**
   * Search jobs
   */
  searchJobs(
  type?: string | null,
  categoryId?: number | null,
  workmode?: string | null,
  experience?: string | null
): Observable<PaginatedResult<Job>> {

  let params = new HttpParams();

  if (type) {
    params = params.set('type', type);
  }

  if (categoryId !== undefined && categoryId !== null) {
    params = params.set('categoryId', categoryId);
  }

  if (workmode) {
    params = params.set('workmode', workmode);
  }

  if (experience) {
    params = params.set('experience', experience);
  }

  return this.http.get<PaginatedResult<Job>>(
    `${this.apiUrl}/search`,
    { params }
  );
}
}

File: JobPortal.Application/Services/JobService.cs
using JobPortal.Application.DTOs;
using JobPortal.Application.Interfaces;
using JobPortal.Application.Interfaces.Services;
using JobPortal.Domain.Entities;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static Microsoft.EntityFrameworkCore.DbLoggerCategory.Database;

namespace JobPortal.Application.Services
{
    public class JobService : IJobService
    {
        private readonly IApplicationDbContext _context;
        private readonly IConfiguration _config;

        public JobService(IApplicationDbContext context, IConfiguration config)
        {
            _context = context;
            _config = config;
        }

        public async Task<bool> CompanyExistsAsync(int id)
        {
            return await _context.Companies.AnyAsync(c => c.CompanyId == id);
        }

        public async Task<JobApplication> CreateJobApplication(CreateApplicationDto dto)
        {
            string? resumePath = null;

            // Save uploaded resume file
            if (dto.ResumeFile != null && dto.ResumeFile.Length > 0)
            {
                var uploadsFolder = Path.Combine(Directory.GetCurrentDirectory(), "wwwroot", "resumes");

                if (!Directory.Exists(uploadsFolder))
                {
                    Directory.CreateDirectory(uploadsFolder);
                }

                var uniqueFileName =
                    $"{Guid.NewGuid()}_{dto.ResumeFile.FileName}";

                var filePath = Path.Combine(uploadsFolder, uniqueFileName);

                using (var stream = new FileStream(filePath, FileMode.Create))
                {
                    await dto.ResumeFile.CopyToAsync(stream);
                }

                resumePath = $"/resumes/{uniqueFileName}";
            }

            var application = new JobApplication
            {
                JobId = dto.JobId,
                CandidateProfileId = dto.CandidateProfileId,
                CoverLetter = dto.CoverLetter,
                ResumeUrl = resumePath,
                Status = "Submitted",
                AppliedAt = DateTime.UtcNow
            };

            _context.Applications.Add(application);

            await _context.SaveChangesAsync();


            return application;
        }

        public async Task<Job> CreateJobAsync(CreateJobDto jobDto)
        {
            var company = await _context.Companies.FindAsync(jobDto.CompanyId); 

            var job = new Job
            {
                CompanyId = jobDto.CompanyId,
                PostedByUserId = jobDto.PostedByUserId,
                CategoryId = jobDto.CategoryId,
                Title = jobDto.Title,
                JobType = jobDto.JobType,
                WorkMode = jobDto.WorkMode,
                Location = jobDto.Location,
                SalaryMin = jobDto.SalaryMin,
                SalaryMax = jobDto.SalaryMax,
                Description = jobDto.Description,
                Responsibilities = jobDto.Responsibilities,
                Requirements = jobDto.Requirements,
                Benefits = jobDto.Benefits,
                Deadline = jobDto.Deadline,
                Status = jobDto.Status,
                Company = company
            };

            _context.Jobs.Add(job);
            await _context.SaveChangesAsync();

            return job;
        }

        public async Task DeleteJobAsync(int id)
        {
            var job = await _context.Jobs.FindAsync(id);
            if (job == null)
            {
                throw new KeyNotFoundException($"Job with ID {id} not found.");
            }

            _context.Jobs.Remove(job);
            await _context.SaveChangesAsync();
        }

        public async Task<Job> GetJobAsync(int id)
        {
            var job = await _context.Jobs
            .Include(j => j.Company)
            .FirstOrDefaultAsync(j => j.JobId == id);

            return job;
        }

        public async Task<Job?> GetJobByIdAsync(int id)
        {
            return await _context.Jobs.FindAsync(id);
        }

        public async Task<IEnumerable<Job>> GetJobsAsync()
        {
            return await _context.Jobs.Include(j=> j.Company).ToListAsync();
        }

        public async Task<PaginatedResult<Job>> GetJobsPaginated(int page = 1, int pageSize = 10)
        {
            var totalJobs = await _context.Jobs.CountAsync();
            var jobs = await _context.Jobs
                 .AsNoTracking()
                .Include(j => j.Company)
                .OrderByDescending(j => j.JobId)
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            var result = new PaginatedResult<Job>
            {
                Items = jobs,
                TotalCount = totalJobs,
                Page = page,
                PageSize = pageSize
            };
          
            return result;
        }

        public async Task<PaginatedResult<Job>> SearchJobs(string? type, int? categoryId, string? workmode, string? experience)
        {
            var query = _context.Jobs.Include(j => j.Company).AsQueryable();

            if (!string.IsNullOrEmpty(workmode))
            {
                query = query.Where(j => j.WorkMode!.Contains(workmode));
            }

            if (!string.IsNullOrEmpty(type))
            {
                query = query.Where(j => type.ToLower().Contains(j.JobType.ToLower()));
            }

            if (categoryId.HasValue)
            {
                query = query.Where(j => j.CategoryId == categoryId.Value);
            }

            var totalJobs = await query.CountAsync();
            var jobs = await query.AsNoTracking()
                .Include(j => j.Company)
                .OrderByDescending(j => j.JobId)
                .Skip(0 * 10)
                .Take(10)
                .ToListAsync();

            var result = new PaginatedResult<Job>
            {
                Items = jobs,
                TotalCount = totalJobs,
                Page = 1,
                PageSize = 10
            };


            return result;
        }

        public async Task<Job?> UpdateJobAsync(int id, UpdateJobDto dto)
        {
            var existingJob = await _context.Jobs.FindAsync(id);
            if (existingJob == null)
            {
                return null;
            }

            existingJob.CompanyId = dto.CompanyId;
            existingJob.PostedByUserId = dto.PostedByUserId;
            existingJob.CategoryId = dto.CategoryId;
            existingJob.Title = dto.Title;
            existingJob.JobType = dto.JobType;
            existingJob.WorkMode = dto.WorkMode;
            existingJob.Location = dto.Location;
            existingJob.SalaryMin = dto.SalaryMin;
            existingJob.SalaryMax = dto.SalaryMax;
            existingJob.Description = dto.Description;
            existingJob.Responsibilities = dto.Responsibilities;
            existingJob.Requirements = dto.Requirements;
            existingJob.Benefits = dto.Benefits;
            existingJob.Deadline = dto.Deadline;
            existingJob.Status = dto.Status;
            existingJob.UpdatedAt = DateTime.UtcNow;

            await _context.SaveChangesAsync();
            return existingJob;
        }
    }
}
